[TypeChecker] Type wrappers (experimental feature)

xedin opened the pull request:
Initial implementation of `@typeWrapper` attribute which could be used to declare a type
 to "wrap" other types by routing all their stored property accesses through itself.

The type wrapper is very similar to a property wrapper but there are couple of fundamental 
differences:

- Could only be associated with a class or a struct;
- Disables memberwise and default initializers, produces a special synthesized initializer that covers all the stored properties instead;
- All access to the stored properties is routed through a single instance of a type wrapper;
- A per-instance `$Storage` container allocates storage for all of the wrapped properties;
- Default initialization expressions are moved to the synthesized constructor;
- Access is performed via subscripts that accept key paths to the underlying storage.

Valid type wrapper has to have:

- A single generic parameter (i.e. `Storage`) which would be substituted with storage type;
- `init(memberwise: Storage)` - initializer that accepts a generic parameter of underlying storage;
- `subscript<Value>(storageKeyPath path: {Writable, ReferenceWritable}KeyPath<Storage, Value>) -> Value` -
  a subscript that would be called on every applicable property access.

Here is an example of a valid type wrapper declaration:

```swift
@typeWrapper
struct Wrapper<Storage> {
  var storage: Storage

  init(memberwise: Storage) {
    self.storage = memberwise 
  }

  subscript<Value>(storageKeyPath path: WritableKeyPath<Storage, Value>) -> Value {
    get { storage[keyPath: path] }
    set { storage[keyPath: path] = newValue }
  }
}
```

Now, let's see `Wrapper` in action to appreciate how much boiler place it removes:

```swift
@Wrapper
struct Person {
  var name: String
  var age: Int
}
```

Access to `name` and `age` of a `Person` is going to get routed through `Wrapper` type.

The transformed type looks like this:

```swift
struct Person {
  struct Storage {
    var name: String
    var age: Int
  }

  var _storage: Wrapper<Storage>

  init(name: String, age: Int) {
    self._storage = Wrapper(memberwise: Storage(name: name, age: age))
  }

  var name: String {
    get { _storage[storageKeyPath: \Storage.name] }
    set { _storage[storageKeyPath: \Storage.name] = newValue }
  }

  var age: Int {
    get { _storage[storageKeyPath: \Storage.age] }
    set { _storage[storageKeyPath: \Storage.age] = newValue }
  }
}
```
All of the boilerplate code required to create and maintain 1-1 storage between type and `Storage` and proper initialization is now handled by the compiler.

Open questions:
  - Support for pre-existing user-defined initializers
  - ✅ Support for stored properties with property wrappers

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin committed:
[AST] Add a new `@typeWrapper` attribute

xedin committed:
[AST] Add a way to check whether type has a type wrapper

xedin committed:
[AST] TypeWrappers: Add a way to inject `$Storage` into a type

`$Storage` type is going to used as a type of `$_storage` variable
and contains all of the stored properties of the type it's attached
to.

xedin committed:
[AST] TypeWrappers: Add a request to inject or get `$_storage` property

This is the property that would get used to route stored property
accesses through a type wrapper.

xedin committed:
[AST] TypeWrappers: Add a request to create and get "mirror" type wrapper property

Given a stored property associated with a type wrapped type,
produce a property that mirrors it in the type wrapper context.

xedin committed:
[TypeChecker] Synthesize getters for stored properties of a type wrapped type

A getter routes accesses through a subscript of `$_storage` variable
injected into a wrapped type.

xedin committed:
[TypeChecker] Synthesize setters for stored properties of a type wrapped type

A setter routes assignment through a subscript of `$_storage` variable
injected into a wrapped type - `$_storage[storageKeyPath: \$Storage.<property>] = newValue`

xedin committed:
[Sema] Don't attempt init synthesis for type wrapped types

Compiler cannot synthesize regular memberwise or default
initializers for type wrapped types because stored properties
of such a type cannot be accessed directly.

A special initializer would be synthesized instead, it is going
to initialize `$_storage` variable and handle default initialization
of stored properties.

xedin committed:
[AST] TypeWrapper: Turn `VarDecl::isAccessibleViaTypeWrapper()` into a cached request

xedin committed:
[TypeChecker] Synthesize initializer for a type wrapped declaration

Synthesize an `init` declaration which is going to initialize
type wrapper instance property `$_storage` via user provided
values for all stored properties.

xedin committed:
[Sema] TypeWrappers: Make sure that type wrapped properties are always "computed"

xedin committed:
[AST] Make it possible to access type wrapper storage of a var

This is important because we need to force existance of the
underlying storage at the right moment.

xedin committed:
[AST] Make it possible to access type wrapper property (storage)

This is important because we need to force existance of the
underlying storage at the right moment.

xedin committed:
[Sema] TypeWrappers: Make sure that synthesized accessors for wrapped vars are not transparent

Type wrapped variables loose their storage and are accessed through
`$Storage` and synthesized accessors cannot be transparent.

xedin committed:
[AST] Add a way to obtain a type of type wrapper

xedin committed:
[TypeChecker] NFC: Add more property wrapper tests

xedin committed:
[Frontend] Mark 'Type Wrappers' as experimental feature that has to be enabled

xedin committed:
[Sema] TypeWrappers: convert variable init expr into initializer default

All of the stored properties are wrapped which means that their
initializers are subsummed and moved to the synthesized `init`
as default arguments.

LucianoPAlmeida commented:
Hey @xedin,
One question, consider this
```swift
@Wrapper
struct Person {
  var name: String
  let age: Int = 0
}
```
Since `wrapper` subscript is defined as `WritableKeyPath` and defines a `set` which cannot be applicable to a `let` property what would be the expected behavior in this case?

xedin commented:
@LucianoPAlmeida properties like that are not currently wrapped.

xedin commented:
A more interesting example is the one which uses a class as a let e.g.:

```swift
class X {
  var v: Int = 42
}

struct S {
  let x: X = X()
}
```

Since `x` is a reference type it would still be modifiable e.g.:

```swift
var s = S()
s.x.v = 0
```

I think this case would have to be wrapped but it would have to pick different overloads of `subscript` from type wrapper type...

LucianoPAlmeida commented:
> I think this case would have to be wrapped but it would have to pick different overloads of `subscript` from type wrapper type...

Yeah, from my understanding an overload of `subscript` that takes an `WritableKeyPath` would not work in that `let reference` case, so user would have to implement a `ReferenceWritableKeyPath` overload of the subscript. 
```swift
let wkp: WritableKeyPath<X, Int> = \.v
let rwkp: ReferenceWritableKeyPath<X, Int> = \.v
let s = S()
s.x[keyPath: wkp] = 0 // error: cannot assign through subscript: 'x' is a let constant
s.x[keyPath: rwkp] = 0 // ok
``` 
Is that what you mean by different overload?

xedin commented:
> Is that what you mean by different overload?

Yes, it's currently modeled as implicitly generated AST so type wrapper could define multiple overloads of subscript and type-checker would just pick one that works for synthesized code.

LucianoPAlmeida commented:
> Yes, it's currently modeled as implicitly generated AST so type wrapper could define multiple overloads of subscript and type-checker would just pick one that works for synthesized code.

Ah got it, Thanks!

xedin committed:
[Sema] TypeWrappers: Allow wrapping stored properties with attached property wrappers

Type wrapper doesn't wrap anything expect to a private backing storage property.
This means that type wrapper is applied first and subsequent `.wrappedValue` and/or
`.projectedValue` is referenced for the returned property wrapper type.

xedin committed:
[Sema] TypeWrappers/NFC: Move initializer synthesis into a request

xedin committed:
[Sema] Allow implicit parameter decls to have property wrapper attributes

xedin committed:
[Sema] TypeWrappers: Synthesize init parameters for property wrapped members

Since properties with property wrappers are not supported, default
init synthesis needs to handle them as well by using correct interface
type depending on outer wrapper capabilities and setting correct
default expression.

xedin commented:
@swift-ci please clean test

xedin commented:
@swift-ci please clean test

xedin commented:
@swift-ci please clean test

xedin committed:
[TypeWrappers] NFC: Fix type wrapper executable test on Linux

xedin commented:
@swift-ci please clean test Linux platform

xedin commented:
@swift-ci please test macOS platform

xedin commented:
@swift-ci please test Windows platform

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test macOS platform

xedin committed:
[Sema] CodeSynthesis: Extract logic to synthesize a parameter for memberwise `init`

xedin committed:
[Sema] TypeWrappers: Add unamanged stored properties to the synthesized memberwise `init`

If a stored property would be in a default memberwise initializer
it would be added to the `init` synthesized for a type wrapped type
as well i.e. a `let` property without a default.

xedin committed:
[Sema] TypeWrappers: Mark `$Storage` struct as `internal`

xedin committed:
[Sema] TypeWrappers: Mark `$_storage` as `private` and all `$Storage` properties as `internal`

xedin commented:
@swift-ci please test
